compiler/il/OMRNode.cpp

```diff
@@ -2357,6 +2357,83 @@ OMR::Node::isNotCollected()
    return false;
    }
 
+/**
+ * \brief
+ *    Check if the node points to the middle of an object/array
+ *
+ * \return
+ *    Return true if the node is pointing to the middle of an object/array, false otherwise
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from a collectable reference, i.e. anything that is an address to a collected object.
+ */
+bool
+OMR::Node::canBeInternalPtrOfObject()
+   {
+   if (!self()->getOpCode().isArrayRef() && !self()->getOpCode().hasPinningArrayPointer())
+      return false;
+
+   TR::Compilation * comp = TR::comp();
+   TR::NodeChecklist visited(comp);
+   TR::Node *curNode = self();
+
+   while (curNode)
+      {
+      if (curNode->isInternalPointer())
+         return true;
+
+      // Come across a node that has already been checked
+      if (visited.contains(curNode))
+         return false;
+
+      if (curNode->getOpCode().hasSymbolReference())
+         {
+         TR::ILOpCode op = curNode->getOpCode();
+         TR::Symbol *symbol = curNode->getSymbolReference()->getSymbol();
+         if (op.isNew() ||
+             (op.getDataType() == TR::Address && op.isCall()) ||
+             symbol->isInternalPointer() ||
+             symbol->isCollectedReference())
+             return true;
+         }
+
+      visited.add(curNode);
+      if (curNode->getNumChildren() >= 1)
```

Comment:
According to previous discussion, there are opcodes this query cannot handle properly, such as aternary. The assertion is to make sure only the cases we expect to handle will be handled. The first if in this function ensures we only deal with aiadd/aladd. This if contains a whitelist of opcodes this query can handle properly.